core/src/main/scala/org/apache/spark/sql/delta/RowId.scala

```diff
@@ -87,4 +90,120 @@ object RowId {
     }
     latestMetadata
   }
+
+  /**
+   * Iterator to assign fresh row IDs to AddFiles in an iterator of actions.
+   */
+  private class RowIdAssigningIterator(
+      actions: Iterator[Action],
+      highWatermark: Option[Long]) extends Iterator[Action] {
+
+    val oldHighWatermark: Long = highWatermark.getOrElse(-1L)
+    var newHighWatermark: Long = oldHighWatermark
+
+    override def hasNext: Boolean = actions.hasNext || (newHighWatermark != oldHighWatermark)
+
+    override def next(): Action = {
+      if (actions.hasNext) {
+        actions.next() match {
+          case a: AddFile if a.baseRowId.isEmpty =>
```

Comment:
is it not a requirement that basedRowId is empty here? Otherwise we are in a re-assign case and have to update?

Reply:
It should always be empty but I don't think we can assume it is since it depends on the caller.
The nice thing with only updating Add files that are missing a baseRowId is that this assignment operation is idempotent.